### ISSUE TITLE
Decide which functions to export

### ISSUE THREAD

**Issue by alecloudenback** [JuliaActuary/Yields.jl#79]

**Comment by kasperrisager** [comment]
Maybe it's because I use C# in my day job, but I would prefer giving concepts long and descriptive names that don't collide with other contexts, and then export them. Like `ContinuousCompounding()` over `Yields.Continuous()`, or `ContinuousRate(0.1)` over `Yields.Continuous(0.1)`. In this case, the long names would also provide a clearer distinction between what is a compounding frequency and what is a rate with a particular compounding frequency.

A general question is, whether `struct`s should be exported. Ideally, they should be implementation details that the user should not worry about, but that would require that `Yields.jl` export some nice helper function that construct them. This is done presently for `YieldCurve`s, but I'm not sure it's universal.

One thing that confuses me is the distinction between the `rate` and `zero` functions. Are there curves where they would meaningfully differ? Should all implementations of `AbstractYieldCurve` be required to provide both? Also, the name `zero` is a little unfortunate, since it is generally used for the additive identity for a type.

**Comment by alecloudenback** [comment]
I'll try and take the time to think more about the points/examples you gave about the names and respond about them when I get more time, but I wanted to quickly answer this part:

> One thing that confuses me is the distinction between the rate and zero functions. Are there curves where they would meaningfully differ? Should all implementations of AbstractYieldCurve be required to provide both? Also, the name zero is a little unfortunate, since it is generally used for the additive identity for a type.

## `zero`

I think I probably chose `zero` because I was referencing "Options, Futures, and Other Derivatives" which preferred the term `zero` over `spot` but I think it may be more common in other places to use `spot`. Do you think that would be preferable?

![image](https://user-images.githubusercontent.com/711879/151476223-b3c558db-f848-4a33-9cf8-8364059fdc10.png)


## `rate`

`rate` is effectively `convert(Float64,rate<:Rate)`, or said in plain english: the way to get the floating point value inside, e.g., `rate(Periodic(0.03,1))` returns `0.03`. 

I'm open to other names or maybe just enforce the convention of using `convert` explicity?

This brings up a related issue, which is that in ActuaryUtilities.jl, a function like `irr` currently returns a `Float64` instead of a `Yields.Periodic`. I've thought about whether it should return a `Rate` type so that it carries the semantic information that it's a periodic rate (instead of, e.g. continuous) and for consistency across the JuliaActuary packages. I haven't felt so strongly about it to initiate that change, but I'm curious what you think?

**Comment by kasperrisager** [comment]
Ah, ok, I get the distinction between `zero` and `rate` then, thanks. I think what confused me was that `rate(::Constant{T})` returns a `T` which is often a `Rate`.

As for naming, I actually think `zero` would be a good name if it hadn't been a semi-reserved word. Maybe `zerorate` or `zero_rate`? I think `spot` could be confusing, e.g, a spot swap (as opposed to forward starting) is still quoted as a par rate.

I think it would be neat to have `irr` and similar return a `Rate`. It would provide the user with a lot of safety to work only with `Rate` instances and not with naturally ambiguous `Float64`s - no accidental mess-ups in compounding convention (as I have done uncountably many times myself in every other language I have ever used - plus Excel 😆).

**Comment by alecloudenback** [comment]
> Ah, ok, I get the distinction between `zero` and `rate` then, thanks. I think what confused me was that `rate(::Constant{T})` returns a `T` which is often a `Rate`.

Do you prefer `rate(r:<Rate)::Float` or enforce the convention of `convert(Float,r:<Rate)`? I lean towards the first unless the name is too confusing.

> As for naming, I actually think `zero` would be a good name if it hadn't been a semi-reserved word. Maybe `zerorate` or `zero_rate`? I think `spot` could be confusing, e.g, a spot swap (as opposed to forward starting) is still quoted as a par rate.

I think `zero` is okay: the only usage in Base/stdlib is a one-argument `zero(x)` whereas Yields only has a multple argument version, e.g. `zero(yield,time)`.

I could see a use for a additive identity element, e.g. if additively reducing over a collection of rates representing yield components (e.g. risk free + credit + liquidity + ...), which makes me think maybe `zero` should be reserved for a yield curve with zero rates. #82

> I think it would be neat to have `irr` and similar return a `Rate`. It would provide the user with a lot of safety to work only with `Rate` instances and not with naturally ambiguous `Float64`s - no accidental mess-ups in compounding convention (as I have done uncountably many times myself in every other language I have ever used - plus Excel 😆).

I agree: https://github.com/JuliaActuary/ActuaryUtilities.jl/issues/62

**Comment by kasperrisager** [comment]
Sorry, I got too many notifications and missed that there was a question here.

> Do you prefer rate(r:<Rate)::Float or enforce the convention of convert(Float,r:<Rate)? I lean towards the first unless the name is too confusing.

I don't have that much of an opinion here. I would be fine with both. I could add in a third possibility which is to not provide these at all. That would make it clear to the user that they are leaving the protected space of typed rates by using `my_rate.value`. But maybe that would just be annoying 😄.

> I think zero is okay: the only usage in Base/stdlib is a one-argument zero(x) whereas Yields only has a multple argument version, e.g. zero(yield,time).

👍

**Comment by alecloudenback** [comment]
>Maybe it's because I use C# in my day job, but I would prefer giving concepts long and descriptive names that don't collide with other contexts, and then export them. Like ContinuousCompounding() over Yields.Continuous(), or ContinuousRate(0.1) over Yields.Continuous(0.1). 

Sorry, I thought I remembered responding to this point but maybe I never submitted it... 

For the naming, I think a pretty common pattern, especially in production code, is to qualify imported functions with the pacakge name to make it clear where a particular function is coming from. 

E.g.

```julia
import Yields

r = Yields.Continuous(0.05)
Yields.discount(r,...)
```

over

```julia
using Yields

r = Continuous(0.05)
discount(r,....)
```

So in this case, having a name like `ContinuousCompounding` would end up being being `Yields.ContinuousCompounding` which is pretty verbose and `Compounding` and `Yields` becomes pretty redundant.

I agree that, on it's own `Continuous` would be a bit vague but I think the package name can provide the needed context and is my preferred solution in the tradeoff between conciseness, lack of ambiguity, and ergonomics. 

The above is attempting to keep the argument self contained and stand on its own without a comparison to other languages, but in comparison to other languages my opinion is that it strikes a balance between the `LongCamelCaseNames` found in some languages and the abstract terseness of, e.g. Haskell.

> In this case, the long names would also provide a clearer distinction between what is a compounding frequency and what is a rate with a particular compounding frequency.

The distinction is really `Rate` vs `CompoundingFrequency` and could be enforced by only allowing a `Rate` construction explicitly: e.g. `Rate(0.05,Continuous())` instead of `Continuous(0.05)` but out of convienince and ergonmoics the former was removed from the exported API (#83) per this suggestion (maybe I misinterpreted).

> A general question is, whether structs should be exported. Ideally, they should be implementation details that the user should not worry about, but that would require that Yields.jl export some nice helper function that construct them. This is done presently for YieldCurves, but I'm not sure it's universal.

In summary: I feel like the names are in the right place and has been pleasant in everyday usage, but maybe that's not representative of other's experience. Please let me know if you feel like that's the case and worth further consideration. Thanks for the continued thoughts.
